obs-webrtc: Add Simulcast Support

Sean-Der opened the pull request:
## Description
This PR adds supports for Simulcast to the obs-webrtc output. Simulcast allows for multiple quality levels to be sent over one track in WebRTC.

[RFC 8853](https://datatracker.ietf.org/doc/rfc8853/) is a technical description. For a less technical explanation [Wowza](https://www.wowza.com/blog/webrtc-simulcast-what-it-is-and-how-it-works), [Dolby](https://docs.dolby.io/streaming-apis/docs/using-webrtc-simulcast) and [LiveKit](https://blog.livekit.io/an-introduction-to-webrtc-simulcast-6c5f1f6402eb/) have all written articles on it.

This PR starts with adding a Spinbox when `WHIP` is selected. This allows the user to choose how many layers to send in total. The acceptable range (for now) is 1-4.
![defaultPage](https://github.com/user-attachments/assets/ed5e3693-9442-4882-bb44-c7553eed3585)

The Height/Width/Bitrate is scaled from the users global choice. So depending on the users choice.
* 1 Layers (
  * 100% Height/Width/Bitrate
* 2 Layers 
  * 100% Height/Width/Bitrate,
  * 50% Height/Width/Bitrate
* 3 Layers 
  * 100% Height/Width/Bitrate
  * 66% Height/Width/Bitrate
  * 33% Height/Width/Bitrate
* 4 Layers 
  * 100% Height/Width/Bitrate
  * 75% Height/Width/Bitrate 
  * 50% Height/Width/Bitrate 
  * 25% Height/Width/Bitrate

A server may wish to only accept a subset of the layers offered. Today if the user rejects any layers we throw an error and say how many layers were accepted. The user can lower the amount of layers they are sending and try again. Services in their documentation will be expected to tell users how they want video encoded and how many layers they are willing to accept.

![failedToAcceptLayer](https://github.com/user-attachments/assets/73106175-4880-4f62-8a0a-cf0fc434d985)

### Motivation and Context

* Low Latency - Decoding/Re-encoding adds too much latency for interactive content
* Make self hosting easier - Running ffmpeg is costly and difficult to scale
* Streamers more control - If streamers can generate transcodes they control the quality 
* E2E Encryption/Prevent Modification- Streamers can distribute keys to viewers and servers can't modify video anymore.

Some questions that have been asked by reviewers already.

**What about Enhanced Broadcasting?** -  Simulcast is part of an existing standard that already sees [a billion minutes per day](https://bloggeek.me/kranky-geek-webrtc-summary-2022/) so it is here to stay and is already deployed massively. I would like to see Enhanced Broadcasting in the JSON blob be able to specify the protocol. When a user starts a enhanced broadcasting session it should be able to switch between RTMP/WebRTC/$x. I think both flows can/should exist because they serve different purposes. 

**UI/UX Considerations** - A single opt-in is non-intrusive and respects the user. It would be good to get feedback from all OBS WebRTC users (Dolby, LiveKit, Cloudflare etc...) that this is a sound decision though. 

### How Has This Been Tested?
This has been tested with the nvenc, qsv, and x264.

I have tested this against Twitch, Broadcast Box, LiveKit and Dolby. 

### Types of changes
- New feature (non-breaking change which adds functionality)


### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

Sean-Der commented:
Hi @tytan652 I see you gave a 👎 was that an accident? Do you have any feedback/things I can improve?

tytan652 commented:
> Hi @tytan652 I see you gave a 👎 was that an accident?

No, I used that to avoid always adding an "off-topic" to say that the PR adds code complexity to the streaming codepath which adds more work to my plans for a Service Overhaul.

> Do you have any feedback/things I can improve?

No, this is why I just reacted.

Sean-Der commented:
I rebased and fixed some merge conflicts. I tested locally again with x264, qsv and nvenc and works. Always appreciate more testing though if anyone wants to try.

thank you

koolscooby commented:
This would be great. We'd like to have Twitch Stream Together users be able to use OBS and use Simulcast to send multiple qualities to improve the participant experience. It's a much better quality encoding & transmission experience compared to in-browser content contribution.

One UI nit / question: does checking the box *force* simulcast, or does it just enable it if the server/SDP exchange also indicates support for simulcast? If the latter, I'd recommend changing the UI label to `Enable simulcast` which makes it a bit clear that it is enabled on the client, but also requires the server to support it.

Sean-Der commented:
@koolscooby Fixed! Changed from `Use` -> `Enabled`

If the server rejects Simulcast it will only accept the high layer.

In the future the server should be able to drive more (how many layers, suggestions around bitrate/resolution). For a V1 I think it just being .5x and .25x is a good trade-off on complexity vs value.

koolscooby commented:
> @koolscooby Fixed! Changed from `Use` -> `Enabled`
> 
> If the server rejects Simulcast it will only accept the high layer.
> 
> In the future the server should be able to drive more (how many layers, suggestions around bitrate/resolution). For a V1 I think it just being .5x and .25x is a good trade-off on complexity vs value.

This seems like a completely sensible tradeoff to start getting users testing the underlying simulcast functionality in OBS & on the server side in a bigger set of network scenarios.

Fenrirthviti commented:
We're aware this is here, and will get to it in due time. More testing/documentation is always welcome to help things.

Sean-Der commented:
@Fenrirthviti I feel very confident with testing.

What can I do to improve documentation? Who is best champion of that among OBS contributors

davidzhao commented:
There are a lot of LiveKit users that are excited about the ability to simulcast with OBS. Let me know if we can help in getting this across the finish line

Fenrirthviti commented:
Logging for posterity that this has an [open RFC](https://github.com/obsproject/rfcs/pull/55) for OBS with unresolved feedback that has not been accepted.

While not impossible this is merged first, it's a little strange to submit a functional PR of a design that has not been accepted by the project or finalized yet. I appreciate the enthusiasm but "move fast, fix later" is at odds with how the project operates in general.

This is not a request for discussion on this point, I am posting for visibility of process.

Sean-Der commented:
What feedback have I not responded too? I had feedback about moving the checkbox and not allowing advanced configuration [here](https://github.com/obsproject/rfcs/pull/55#issuecomment-1627791991) that I responded to over a year ago.

Should I be engaging with @Warchamp7 differently?

thanks

nils-ohlmeier commented:
We as Cloudflare are in support of adding Simulcast support to OBS. Having Simulcast support in OBS will enable new use cases for certain users and services.
I think the default values chosen in this PR for 1/2 and 1/4 resolution are used by a lot of video conferencing services in their Simulcast settings and make perfect sense as the initial values for this feature.

kwindla commented:
I'd like to weigh in to say that we (Daily.co) have users who need this. In general, our experience is that simulcast is critical for expanding the use cases and kinds of users who can stream successfully. Two examples: 1) users who don't have control over their upstream bandwidth. Perhaps they are on a corporate network with unpredictable congestion. 2) users who want to live stream and send the stream out to multiple different protocols/channels in real-time without re-processing.

Sean-Der commented:
I have had conversations with OBS Maintainers, WHIP Services and the WHIP spec author about Simulcast and have updated the PR from that feedback. We want to get an MVP into OBS Studio so we can build towards the next/future items and would appreciate another round of review now that I’ve refactored the implementation to follow the MVP.

(MVP) - Users are given a Spinbox to select how many Simulcast layers to send. If the remote server doesn't accept the configuration throw an error modal. A screenshot of this is show in the PR description. 

(Next) - Allow servers to configure/hint layers even before the user hits 'Start Streaming'. Users don't need to configure anything. The behavior would feel closer to Twitch's Enhanced Broadcasting.

(Future) Implement dynamic configuration. OBS could enable/disable layers on the fly depending on CPU/Network usage.

Bleuzen commented:
In this current form this is not compatible with other rate control settings than CBR.
For example in VBR mode it only changes resolution and bitrate for each layer but the max bitrate stays the same for every layer which doesn't make sense.

So for this first version I suggest to either only allow users to enable Simulcast in CBR mode or add some special handling for other rate control modes (at least VBR makes sense because in my experience some browsers don't like the bitstream padding from CBR so VBR is more stable for WebRTC streams.)

For the future it would be cool to have entirely separate settings per layer like. I'm not a fan of the calculated resolutions & bitrates here and would rather choose my own and also do things like having one layer with AV1 and another layer with H264. But that is something for the future / another PR I guess.

Sean-Der commented:
@Bleuzen those are all great enhancements. In the future I would like to see

* Dynamic toggling of tracks (don’t generate if no one is viewing)
* Multi codec
* User selection of resolution/bitrate

It’s overwhelmingly complicated though. I hope we can keep this PR as is and just get it merged. If people find it useful I would love to do all this :)

DDRBoxman commented on UI/forms/OBSBasicSettings.ui:
```diff
@@ -1849,6 +1849,91 @@
+                   <item row="1" column="1">
+                    <layout class="QHBoxLayout" name="horizontalLayout_34" stretch="0,0">
+                     <item>
+                      <widget class="QSpinBox" name="simulcastTotalLayers">
```
Looks like the encoding size / bitrate information box from the simulcast RFC is missing

Sean-Der replied:
@DDRBoxman I forgot to remove this from RFC. I was told this is too busy/too much UI.

Sean-Der replied:
Updated RFC

DDRBoxman replied:
I think that this ends up with us back in a similar situation to the opaque checkbox we didn't like in the first place.

Can we add a link below there that pops up a dialog that has static text to explain how the different layers will be calculated and what else to expect.

DDRBoxman commented on UI/data/locale/en-US.ini:
```diff
@@ -985,6 +985,9 @@ Basic.Settings.Stream.MultitrackVideoStreamDumpEnable="Enable stream dump to FLV
 Basic.Settings.Stream.MultitrackVideoConfigOverrideEnable="Enable Config Override"
 Basic.Settings.Stream.MultitrackVideoLabel="Multitrack Video"
+Basic.Settings.Stream.SimulcastLabel="Simulcast"
+Basic.Settings.Stream.SimulcastInfo="Simulcast allows you to encode and send multiple video qualities. No information about your computer and software setup will be sent."
```
We can probably lose the extra shade here. It doesn't tell the user about what the feature does

Sean-Der replied:
@DDRBoxman This is sincere. If users only know RTMP I would like them to know it different.

They control the settings and shouldn’t expect the user to make any changes.

Fenrirthviti replied:
Easiest thing to do is just remove the extra sentence. Debatable that it could describe that it may increase system resources or something else useful to the end-user enabling it, but that can be adjusted/added later if it comes up.

Sean-Der replied:
@Fenrirthviti Done! Second sentence is deleted

DDRBoxman commented on UI/forms/OBSBasicSettings.ui:
```diff
@@ -1849,6 +1849,91 @@
+                    </spacer>
+                   </item>
+                   <item>
+                    <widget class="QLabel" name="simulcastInfo">
```
nit: it might be useful to put all of simulcast under a toggle. That way users don't have to think about an advanced setting even if the default is only the 1 layer.

Sean-Der replied:
I don't see this as an advanced setting. If you are streaming you are probably going to set this to more then 1 (because your streaming provider suggests it)

Sean-Der replied:
So the default flow is going to have an additional click.

DDRBoxman commented on plugins/obs-webrtc/whip-output.cpp:
```diff
@@ -465,31 +535,41 @@ bool WHIPOutput::Connect()
+		auto layersAccepted = CountSimulcastLayers(response);
+		if (videoLayerStates.size() != layersAccepted) {
+			do_log(LOG_ERROR, "WHIP only accepted %lu layers", layersAccepted);
+			displayError(std::to_string(layersAccepted).c_str(), "Error.SimulcastLayersRejected");
```
cool 👍

Sean-Der replied:
Thanks! It is so simple, and I anticipate it will help a lot :)

DDRBoxman commented on plugins/obs-webrtc/whip-output.cpp:
```diff
@@ -321,6 +363,28 @@ void WHIPOutput::ParseLinkHeader(std::string val, std::vector<rtc::IceServer> &i
 	}
 }
 
+size_t WHIPOutput::CountSimulcastLayers(std::string answer)
```
Could probably move this out into a static method since it's just works on its own and doesn't use any class state.

Sean-Der replied:
Fixed. This is now static

DDRBoxman replied:
Might be worth pulling into the utils and fixing up the name to reflect that this is parsing the response from the server. The current naming would have me think it's getting the layers from the output.

Sean-Der replied:
👍 moved into whip-utils.

DDRBoxman commented on UI/window-basic-main-outputs.cpp:
Instead of sprinkling the simulcast throughout this file we should pull out the logic elsewhere.

For example the multi encoder stuff setups hooks and does the building in a separate class:
https://github.com/obsproject/obs-studio/commit/c8950900c36fd6d0bd014eb6fd24da829a63033d#diff-cbc8676187d8579f6909bb6970d56123b2b77250c938b825bc16b89b909cd1a9R725

Sean-Der replied:
I attempted this and wasn't able to reduce the complexity. Multitrack does sprinkles in the same places. Here is one of the places.

I could move the big Simulcast methods out! Would that help?

```
                  for (size_t i = 0; i < simulcastEncoders.size(); i++)
                          obs_output_set_video_encoder2(streamOutput, simulcastEncoders[i], i + 1);
                  obs_output_set_audio_encoder(streamOutput, streamAudioEnc, 0);

                  if (!is_multitrack_output) {
                          obs_output_set_audio_encoder(streamOutput, streamAudioEnc, 0);
                  } else {
```

DDRBoxman replied:
Hmm, I think that check should just be for multiple audio encoders and that supports a few other protocols.


See here where multitrack creates it's encoders Instead of being inside the main output ui file:
https://github.com/obsproject/obs-studio/blob/a0e4e3734126a7e76a871fa671be952de5c13dd8/UI/multitrack-video-output.cpp#L592

Definitely seems like more of this multitrack code could get pulled out of this file too but encapsulating as much of the simulcast stuff as possible would be ideal.

Sean-Der replied:
Would this be the right way?

* Create a class for Simulcast
* Move two simulcast methods into it
* Replace calls to that instead

Multitrack encoder creation happens at a different time compared to Simulcast. So the hooks end up being different

Sean-Der commented:
@DDRBoxman I have moved Simulcast into a dedicated file. Is that a better feel?

DDRBoxman commented:
Yeah having that code in the new file does feel cleaner 👍 

Since it's not applicable to not whip it might be worth changing the naming in window-basic-settings-stream and window-basic-main-outputs from just `simulcast*` to something more specific like *whipSimulcast*' so that future adventurers don't think that it's a more generic feature.

Aside from that I left a comment about having some other ui to at least explain what is in the layers.

Sean-Der commented:
@DDRBoxman Updated everything to be `whipSimulcast` instead of `simulcast`

Could we defer on the UI? I will schedule a 30 min meeting and we can go over that part again. Discussing UI over GitHub/Discord I have a hard time communicating details.

DDRBoxman commented:
We should see if we can get @Warchamp7 on the same call so we're all on the same page

Sean-Der commented:
@DDRBoxman I sent a DM on discord. 

Should I do the static text? Do you have an example that I should clone/copy in OBS already? I would love to just knock it out. Then I can start giving out builds!

 I am just worried that if I make UI/UX changes that didn't get discussed I will get myself in trouble again

DDRBoxman commented:
Since the client will error if the number of layers doesn't match we're good on that front. 

Warchamp suggested we could make a new OBS knowledge base page for this that we can link to. (not sure what the process to add one is but should be easy if we have the copy ready)

I was thinking that just adding a QLabel with an anchor tag that links there would be fine

Sean-Der commented:
Great! I will figure that out and make the page, and update PR.

thanks for getting the ‘next step’ on that :)

Sean-Der commented:
I created a wiki page here https://github.com/obsproject/obs-studio/wiki/Streaming-With-WHIP#simulcast and linked to it.

Could I get another review, thank you!

DDRBoxman commented:
Yeah the href seems like a fine way to do that. I think this is supposed to be in the knowledge base and not the wiki though since that's more developer focused.

Sean-Der commented:
@DDRBoxman I am not able to do that, do you have access to it? 

Once it is up I can change link right away!

DDRBoxman commented:
I think you'll have to ask in the Discord I don't actually know anything about the knowledge base system

Sean-Der commented:
@DDRBoxman done! 

@Fenrirthviti offered to update it https://discord.com/channels/348973006581923840/939687060162363452/1327845941780156459

When that is done I will update and request another review, thanks again for all the help

Sean-Der commented:
@DDRBoxman PR has been updated with official URL. I think I got it this time :)

thank you @Fenrirthviti

DDRBoxman reviewed (approved):
Cool, that covers the things that are on my radar 👍

Sean-Der committed:
obs-webrtc: Add Simulcast Support

Sean-Der commented:
I have updated/rebase the PR for the frontend rewrite

getroot commented:
This works well with OvenMediaEngine. https://github.com/AirenSoft/OvenMediaEngine/discussions/1769
I confirmed that when broadcasting to OvenMediaEngine using OBS WHIP Simulcast, it plays well with WebRTC and LLHLS ABR. Thank you for the great feature. I hope this will be merged soon.

mpisat commented:
We are all waiting for this to be merged. It works really well with OvenMediaEngine.
I hope this gets prioritized soon.

Sean-Der commented:
I just rebased and fixed merge conflicts, @mpisat should be good to go! Sorry that took so long :(

mpisat commented:
Will download artifacts and test out. Thank you!

naikrovek commented:
> For the future it would be cool to have entirely separate settings per layer. I'm not a fan of the calculated resolutions & bitrates here and would rather choose my own and also do things like having one layer with AV1 and another layer with H264. But that is something for the future / another PR I guess.

I agree with this _completely_.  More control over the layers would be ideal.

dascope commented:
Awesome, we are also excited and waiting for this since we supported WHIP support in Digital Samba 611.

IsaiahJTurner commented:
Excited for the current implementation to merge, this will definitely meet our needs to stream HIGH, MEDIUM, and LOW layers to LiveKit! Long term, additional customization sounds like a great improvement.

mpisat commented:
> Excited for the current implementation to merge, this will definitely meet our needs to stream HIGH, MEDIUM, and LOW layers to LiveKit! Long term, additional customization sounds like a great improvement.
> 
> Also, found this PR after A LOT of digging to understand where the Simulcast option that is included in the documentation was at.

Livekit Ingress doesn't support simulcast no-transcoding bypass with this OBS. I tested it, it recognizes the layers but layer switching doesn't work. LiveKit needs to modify code to make that happen. I tried to fix myself but I failed. 

However, if this goes mainstream, and merged to OBS, they might fix it. It'd be amazing.
With current obs-simulcast implementation, OvenmediaEngine works pretty awesome (kudos to Ovenmedia team)

Fenrirthviti commented:
This is currently being targeted for 32.1, but requires final review and approvals still which have not yet been completed.

Sean-Der commented:
@peerasan You can use builds from GitHub actions, do those work for you? If you have any problems always here to help :)

Warchamp7 reviewed (changes requested):
Couple minor things to fix, but otherwise this follows all of the suggestions and guidance that have been provided off-thread for an MVP on the UX side.

peerasan commented:
From OBS's logs, I noticed some values might not be correct
- bitrate of whip_simulcast_2 (480x852), whip_simulcast_1 (240x426)  is 6000  Kbps?
- keyint: 250

````bash
12:32:54.485: ---------------------------------
12:32:54.485: [x264 encoder: 'advanced_video_stream'] preset: veryfast
12:32:54.485: [x264 encoder: 'advanced_video_stream'] profile: baseline
12:32:54.485: [x264 encoder: 'advanced_video_stream'] settings:
12:32:54.485: 	rate_control: CBR
12:32:54.485: 	bitrate:      2000
12:32:54.485: 	buffer size:  2000
12:32:54.485: 	crf:          23
12:32:54.485: 	fps_num:      25
12:32:54.485: 	fps_den:      1
12:32:54.485: 	width:        720
12:32:54.485: 	height:       1280
12:32:54.485: 	keyint:       50
12:32:54.485: 
12:32:54.485: [x264 encoder: 'advanced_video_stream'] custom settings: 
12:32:54.485: 	bframes = 0
12:32:54.491: NV12 texture support enabled
12:32:54.491: P010 texture support not available
12:32:54.492: ---------------------------------
12:32:54.492: [x264 encoder: 'whip_simulcast_2'] preset: veryfast
12:32:54.492: [x264 encoder: 'whip_simulcast_2'] settings:
12:32:54.492: 	rate_control: CBR
12:32:54.492: 	bitrate:      6000
12:32:54.492: 	buffer size:  6000
12:32:54.492: 	crf:          23
12:32:54.492: 	fps_num:      25
12:32:54.492: 	fps_den:      1
12:32:54.492: 	width:        480
12:32:54.492: 	height:       852
12:32:54.492: 	keyint:       250
12:32:54.492: 
12:32:54.494: NV12 texture support enabled
12:32:54.494: P010 texture support not available
12:32:54.494: ---------------------------------
12:32:54.494: [x264 encoder: 'whip_simulcast_1'] preset: veryfast
12:32:54.494: [x264 encoder: 'whip_simulcast_1'] settings:
12:32:54.494: 	rate_control: CBR
12:32:54.494: 	bitrate:      6000
12:32:54.494: 	buffer size:  6000
12:32:54.494: 	crf:          23
12:32:54.494: 	fps_num:      25
12:32:54.494: 	fps_den:      1
12:32:54.494: 	width:        240
12:32:54.494: 	height:       426
12:32:54.494: 	keyint:       250
12:32:54.494: 
12:32:54.496: ---------------------------------
```

Sean-Der commented:
@peerasan When I do Simple

```
info: [x264 encoder: 'simple_video_stream'] preset: veryfast
info: [x264 encoder: 'simple_video_stream'] settings:
	rate_control: CBR
	bitrate:      6000
	buffer size:  6000
	crf:          23
	fps_num:      30
	fps_den:      1
	width:        1280
	height:       720
	keyint:       250

info: NV12 texture support enabled
info: P010 texture support not available
info: ---------------------------------
info: [x264 encoder: 'whip_simulcast_3'] preset: veryfast
info: [x264 encoder: 'whip_simulcast_3'] settings:
	rate_control: CBR
	bitrate:      4500
	buffer size:  4500
	crf:          23
	fps_num:      30
	fps_den:      1
	width:        960
	height:       540
	keyint:       250

info: NV12 texture support enabled
info: P010 texture support not available
info: ---------------------------------
info: [x264 encoder: 'whip_simulcast_2'] preset: veryfast
info: [x264 encoder: 'whip_simulcast_2'] settings:
	rate_control: CBR
	bitrate:      3000
	buffer size:  3000
	crf:          23
	fps_num:      30
	fps_den:      1
	width:        640
	height:       360
	keyint:       250

info: NV12 texture support enabled
info: P010 texture support not available
info: ---------------------------------
info: [x264 encoder: 'whip_simulcast_1'] preset: veryfast
info: [x264 encoder: 'whip_simulcast_1'] settings:
	rate_control: CBR
	bitrate:      1500
	buffer size:  1500
	crf:          23
	fps_num:      30
	fps_den:      1
	width:        320
	height:       180
	keyint:       250
```

When I do advanced
```

info: [x264 encoder: 'advanced_video_stream'] preset: veryfast
info: [x264 encoder: 'advanced_video_stream'] settings:
	rate_control: CBR
	bitrate:      6000
	buffer size:  6000
	crf:          23
	fps_num:      30
	fps_den:      1
	width:        1280
	height:       720
	keyint:       250

info: NV12 texture support enabled
info: P010 texture support not available
info: ---------------------------------
info: [x264 encoder: 'whip_simulcast_3'] preset: veryfast
info: [x264 encoder: 'whip_simulcast_3'] settings:
	rate_control: CBR
	bitrate:      4500
	buffer size:  4500
	crf:          23
	fps_num:      30
	fps_den:      1
	width:        960
	height:       540
	keyint:       250

info: NV12 texture support enabled
info: P010 texture support not available
info: ---------------------------------
info: [x264 encoder: 'whip_simulcast_2'] preset: veryfast
info: [x264 encoder: 'whip_simulcast_2'] settings:
	rate_control: CBR
	bitrate:      3000
	buffer size:  3000
	crf:          23
	fps_num:      30
	fps_den:      1
	width:        640
	height:       360
	keyint:       250

info: NV12 texture support enabled
info: P010 texture support not available
info: ---------------------------------
info: [x264 encoder: 'whip_simulcast_1'] preset: veryfast
info: [x264 encoder: 'whip_simulcast_1'] settings:
	rate_control: CBR
	bitrate:      1500
	buffer size:  1500
	crf:          23
	fps_num:      30
	fps_den:      1
	width:        320
	height:       180
	keyint:       250

info: ---------------------------------
info: [FFmpeg libopus encoder: 'adv_stream_audio'] bitrate: 160, channels: 2, channel_layout: stereo, track: 1

debug: [obs-webrtc] [whip_output: 'adv_stream'] WHIP Resource URL is: https://b.siobud.com/api/whip
info: [obs-webrtc] [whip_output: 'adv_stream'] PeerConnection state is now: Connecting
debug: Maximum interleaver batch size for 'adv_stream' calculated to be 11 packets
info: ==== Streaming Start ===============================================
info: [obs-webrtc] [whip_output: 'adv_stream'] PeerConnection state is now: Connected
info: [obs-webrtc] [whip_output: 'adv_stream'] Connect time: 174ms
```

Sean-Der commented:
@peerasan Can you backup your OBS settings, and then try from a fresh config? Any other hints/ideas on reproducing.

RytoEX reviewed (changes requested):
Just a few items.

Sean-Der commented on plugins/obs-webrtc/whip-output.cpp:
```diff
@@ -59,6 +61,18 @@ bool WHIPOutput::Start()
+		}
+
+		auto v = std::make_shared<videoLayerState>();
+		v->ssrc = base_ssrc + 1 + idx;
```
`base_ssrc` is for the audio track. 

We are doing `+1` so we get a unique SSRC (id) for the initial video track.

Sean-Der commented on plugins/obs-webrtc/whip-output.cpp:
```diff
@@ -368,32 +419,41 @@ bool WHIPOutput::Connect()
 	curl_easy_setopt(c, CURLOPT_ERRORBUFFER, error_buffer);
 
-	auto cleanup = [&]() {
+	auto cleanup = [&](bool connectFailed) {
```
Fixed! Change it all to `doCleanup`

Sean-Der commented on plugins/obs-webrtc/whip-output.cpp:
```diff
@@ -475,31 +532,41 @@ bool WHIPOutput::Connect()
 	response.erase(0, response.find("v=0"));
 
+	// If we are sending multiple layers assert
+	// that the remote accepted them all
```
Fixed! Unwrapped

Sean-Der commented:
Seems like flatpak is failing for every PR, not WHIP specific!

RytoEX reviewed (approved):
Looks okay at a glance.

nab-os reviewed (commented):
Looks good to me

Sean-Der commented:
The RIDs are index 0 https://github.com/obsproject/obs-studio/blob/master/plugins/obs-webrtc/whip-output.cpp#L73

It doesn't support `max-br` yet. I could add that though! At signaling time throw an error if the configured bitrate on OBS side exceeds what server supports.

AdelKS commented:
Shouldn't the bitrate scaling follow the amount of pixels instead ? That would reduce the overall used bandwidth with a more coherent quality across streams :thinking: .
